FHEM/00_SIGNALduino.pm

```diff
@@ -1324,11 +1324,17 @@ sub SIGNALduino_CheckVersionResp
 	}
 
 	if (!defined($hash->{version}) ) {
-		$msg = "$name: CheckVersionResp, Not an SIGNALduino device, got for V: $msg";
-		$hash->{logMethod}->($hash, 1, $msg);
-		readingsSingleUpdate($hash, "state", "no SIGNALduino found", 1); #uncoverable statement because state is overwritten by SIGNALduino_CloseDevice
- 		$hash->{DevState} = 'INACTIVE';
-		SIGNALduino_CloseDevice($hash);
+		if ($hash->{initretry} <= SDUINO_INIT_MAXRETRY) {
```

Comment:
Ich bin am überlegen, ob die Abfrage auf SDUINO_INIT_MAXRETRY hier nicht entfallen kann, da dies ja bereits in SIGNALduino_StartInit implementiert ist.

Wie hast Du den Fall denn zwecks Test herbeigeführt dass mehrfach der Initalisierungsprozess versucht wird?

Reply:
Ich bin davon ausgegangen, das die sub SIGNALduino_CheckVersionResp die bisherige sub SIGNALduino_CheckCmdResp ersetzt, deshalb habe ich es dort eingefügt.
Die beiden Subs SIGNALduino_CheckVersionResp und SIGNALduino_StartInit rufen sich ja praktisch gegenseitig auf, bis SDUINO_INIT_MAXRETRY überschritten wird. Dann erfolgt noch ein einmaliger RESET und wenn der auch nicht erfolgreich ist, wird der Port geschlossen.

Ich habe einen SIGNALduino, der an einem ESPEasy hängt. Die serielle Schnittstelle wird dort mittels SoftwareSerial bedient. Das funktioniert allerdings nicht so 100%ig. Deshalb sind meistens mehrere Abfragen der Version erforderlich. @HomeAutoUser hatte das auch schon mal im Issue https://github.com/RFD-FHEM/RFFHEM/issues/795 unter Punkt 4 mit erwähnt.

Reply:
Prinzipiell richig.
SIGNALduino_CheckVersionResp ersetzt SIGNALduino_CheckCmdResp was darauf hindeutet, dass der Spaß schon mal funktioniert hat :)

In SIGNALduino_CheckVersionResp müssen wir meiner Einschätzung nach aber weder ein close noch SDUINO_INIT_MAXRETRY  prüfen, weil das in SIGNALduino_StartInit steckt.

Deinen Aufbau kann ich nicht so auf die schnelle nachbauen.
Wenn ich versuche ihn zu initalisieren und der SIGNALduino nicht erreichbar ist, dann müssten sich die Wiederholungen im Log bemerkbar machen.
3 Wiederholungen und dann noch einmal Reset

Reply:
Ich habe das mit einer angepassten FW mal nachgestellt.

So sieht es ohne deine Anpassung aus. Ich denke das ist der Fehler. Richtig?
```
2020.03.27 01:18:47.270 2: ipDuino: CloseDevice, closed
2020.03.27 01:18:47.256 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 01:18:47.255 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 01:18:37.250 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP cc1101 (chip CC110 unknown) - compiled at Mar 27 2020 01:17:56
2020.03.27 01:18:37.249 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP cc1101 (chip CC110 unknown) - compiled at Mar 27 2020 01:17:56
2020.03.27 01:18:37.248 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP cc1101 (chip CC110 unknown) - compiled at Mar 27 2020 01:17:56
2020.03.27 01:18:37.233 5: ipDuino: SimpleWrite, V
2020.03.27 01:18:37.232 3: ipDuino: StartInit, get version, retry = 0
2020.03.27 01:18:36.734 5: ipDuino: SimpleWrite, XQ
2020.03.27 01:18:36.733 3: ipDuino: SimpleWrite_XQ, disable receiver (XQ)
2020.03.27 01:18:35.228 1: 10.2.11.41:23 reappeared (ipDuino)

```

Reply:
Genau so sah es vor der Anpassung aus und danach so:

```
2020.03.27 07:41:34.768 3: Opening sduinoEasy868 device 192.168.0.32:2323
2020.03.27 07:41:34.769 3: sduinoEasy868: Attr, setting Verbose to: 3
2020.03.27 07:41:36.690 3: sduinoEasy868: IdList, attr whitelist: 0,0.1,0.2,0.3,0.4,1,3,3.1,4,6,7,8,9,10,11,12,13,13.1,13.2,14,15,16,17,17.1,18,20,21,22,23,24,25,26,27,28,29,30,32,33,33.1,33.2,34,35,36,37,38,39,40,41,42,43,44,44.1,45,46,47,48,49,49.1,49.2,50,51,52,53,54,54.1,55,56,57,58,59,60,61,62,64,65,66,67,68,69,70,71,72,73,74,74.1,76,79,80,81,83,84,85,86,87,88,89,90,91,91.1,92,93,94,95,96,97
2020.03.27 07:41:36.691 3: sduinoEasy868: IdList, MS 0 0.1 0.2 0.3 0.4 1 3 3.1 4 6 7 13 13.2 14 15 17 20 23 25 33 33.1 33.2 35 41 49 51 53 54.1 55 65 68 74.1 87 88 90 91.1 93
2020.03.27 07:41:36.691 3: sduinoEasy868: IdList, MU 8 9 13.1 16 17.1 21 22 24 26 27 28 29 30 32 34 36 37 38 39 40 42 44 44.1 45 46 48 49.1 49.2 50 54 56 59 60 61 62 64 66 67 69 70 71 72 73 74 76 79 80 81 83 84 85 86 89 91 92 94 95 97
2020.03.27 07:41:36.691 3: sduinoEasy868: IdList, MC 10 11 12 18 43 47 52 57 58 96
2020.03.27 07:41:36.857 1: sduinoEasy868: DoInit, 192.168.0.32:2323
2020.03.27 07:41:36.858 1: sduinoEasy868: DoInit, 192.168.0.32:2323
2020.03.27 07:41:36.858 3: sduinoEasy868 device opened
2020.03.27 07:41:38.369 3: sduinoEasy868: SimpleWrite_XQ, disable receiver (XQ)
2020.03.27 07:41:38.869 3: sduinoEasy868: StartInit, get version, retry = 0
2020.03.27 07:41:48.881 3: sduinoEasy868: StartInit, get version, retry = 1
2020.03.27 07:41:49.182 2: sduinoEasy868: CheckVersionResp, initialized v3.4.2_dev_24.03
2020.03.27 07:41:49.192 3: sduinoEasy868: CheckVersionResp, enable receiver (XE) 
2020.03.27 07:42:00.894 3: sduinoEasy868: CheckCcpatableResponse, patable: 84
2020.03.27 07:48:49.199 3: sduinoEasy868: KeepAlive, not ok, retry = 2 -> get ping
2020.03.27 07:49:49.200 3: sduinoEasy868: KeepAlive, not ok, retry = 3 -> get ping
2020.03.27 08:04:49.220 3: sduinoEasy868: KeepAlive, not ok, retry = 2 -> get ping
2020.03.27 08:05:49.222 3: sduinoEasy868: KeepAlive, not ok, retry = 3 -> get ping
2020.03.27 08:06:49.223 3: sduinoEasy868: KeepAlive, not ok, retry count reached. Reset
2020.03.27 08:06:49.223 3: sduinoEasy868: ResetDevice, 
2020.03.27 08:06:49.224 3: Opening sduinoEasy868 device 192.168.0.32:2323
2020.03.27 08:06:49.228 1: sduinoEasy868: DoInit, 192.168.0.32:2323
2020.03.27 08:06:49.228 1: sduinoEasy868: DoInit, 192.168.0.32:2323
2020.03.27 08:06:49.228 3: sduinoEasy868 device opened
2020.03.27 08:06:50.730 3: sduinoEasy868: SimpleWrite_XQ, disable receiver (XQ)
2020.03.27 08:06:51.230 3: sduinoEasy868: StartInit, get version, retry = 0
2020.03.27 08:06:51.381 2: sduinoEasy868: CheckVersionResp, initialized v3.4.2_dev_24.03
2020.03.27 08:06:51.392 3: sduinoEasy868: CheckVersionResp, enable receiver (XE) 
2020.03.27 08:17:51.411 3: sduinoEasy868: KeepAlive, not ok, retry = 2 -> get ping
```

Könnten wir eigentlich auch noch eine der beiden DoInit-Ausgaben entfernen? Mir erschließt sich der Sinn der zweiten Ausgabe nicht so richtig.

Reply:
@elektron-bbs 

Ich habe mir die Ausgabe angesehen. Die war ja im Code doppelt:
https://github.com/RFD-FHEM/RFFHEM/commit/c22d89aea2cb374132c6a93f76c082316806cfcb

Das Prüfen von SDUINO_INIT_MAXRETRY habe ich mal entfernt. Funktioniert auf den 1. Blick einwandfrei. Ich muss es aber noch ein paar mal testen.

Reply:
Ich habe für den Test die Firmware so modifiziert, dass sie keinen korrekten Versionsstring sendet.
So ließe sich das Verhalten auch einigermaßen in einem Test verifizieren.

Das ist der Erste Abschnitt, es wird 3x versucht die Version zu erfragen. Danach kommt der einmalige Reset Versuch:
```
2020.03.27 18:48:48.252 3: ipDuino: ResetDevice, 
2020.03.27 18:48:48.252 2: ipDuino: StartInit, retry count reached. Reset
2020.03.27 18:48:48.251 3: ipDuino: StartInit, get version, retry = 3
2020.03.27 18:48:48.235 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:48:48.234 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:48:38.318 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:38.317 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:38.316 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:38.214 5: ipDuino: SimpleWrite, V
2020.03.27 18:48:38.214 3: ipDuino: StartInit, get version, retry = 2
2020.03.27 18:48:38.202 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:48:38.201 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:48:35.923 3: sduino: CheckCcpatableResponse, patable: 84
2020.03.27 18:48:28.280 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:28.279 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:28.278 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:28.183 5: ipDuino: SimpleWrite, V
2020.03.27 18:48:28.182 3: ipDuino: StartInit, get version, retry = 1
2020.03.27 18:48:28.173 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:48:28.173 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:48:26.582 1: Including fhem.cfg
2020.03.27 18:48:24.031 3: sduino: CheckVersionResp, enable receiver (XE) 
2020.03.27 18:48:24.015 2: sduino: CheckVersionResp, initialized v3.4.2_dev_24.03
2020.03.27 18:48:23.725 1: sduino: SD_BELL_Parse UNDEFINED BELL detected, Protocol 42 code FAB029D
2020.03.27 18:48:21.426 3: sduino: StartInit, get version, retry = 1
2020.03.27 18:48:21.407 1: sduino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:48:19.163 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:19.163 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:19.162 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:18.159 5: ipDuino: SimpleWrite, V
2020.03.27 18:48:18.158 3: ipDuino: StartInit, get version, retry = 0
```

Danach gibt es drei neue Versuche mit anschließendem Close, weil es nicht geklappt hat.
```
2020.03.27 18:49:20.462 2: ipDuino: CloseDevice, closed
2020.03.27 18:49:20.462 2: ipDuino: StartInit, init retry count reached. Closed
2020.03.27 18:49:20.461 3: ipDuino: StartInit, get version, retry = 3
2020.03.27 18:49:20.445 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:49:20.444 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:49:10.478 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:10.477 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:10.476 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:10.427 5: ipDuino: SimpleWrite, V
2020.03.27 18:49:10.426 3: ipDuino: StartInit, get version, retry = 2
2020.03.27 18:49:10.410 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:49:10.409 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:49:00.457 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:00.452 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:00.451 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:49:00.392 5: ipDuino: SimpleWrite, V
2020.03.27 18:49:00.391 3: ipDuino: StartInit, get version, retry = 1
2020.03.27 18:49:00.377 1: ipDuino: CheckVersionResp, Not an SIGNALduino device, got for V: undef
2020.03.27 18:49:00.373 5: ipDuino: CheckVersionResp, called without msg
2020.03.27 18:48:50.411 5: ipDuino: Read, msg: regexp=V\s.*SIGNAL(?:duino|ESP).*(?:\s\d\d:\d\d:\d\d) cmd=version msg=DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:50.411 5: ipDuino: Parse, noMsg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:50.410 4: ipDuino: Read, msg: DMMY 3.4.0-dev+20200216 SIGNALESP  - compiled at Mar 27 2020 18:32:56
2020.03.27 18:48:50.356 5: ipDuino: SimpleWrite, V
2020.03.27 18:48:50.356 3: ipDuino: StartInit, get version, retry = 0
2020.03.27 18:48:49.857 5: ipDuino: SimpleWrite, XQ
2020.03.27 18:48:49.856 3: ipDuino: SimpleWrite_XQ, disable receiver (XQ)
2020.03.27 18:48:48.351 3: ipDuino device opened
2020.03.27 18:48:48.350 1: ipDuino: DoInit, 10.2.11.41:23
2020.03.27 18:48:48.257 4: IP: 10.2.11.41 -> 10.2.11.41
2020.03.27 18:48:48.256 5: HttpUtils url=http://10.2.11.41:23/
2020.03.27 18:48:48.254 3: Opening ipDuino device 10.2.11.41:23
```


Sollte so alles richtig sein oder?

Reply:
Soll ich die angepasste Version hier committen oder passt damit irgendwas nicht, was ich übersehen habe.

Reply:
Ich denke mal ja. Einen Test auf dem betroffenen System kann ich erst am Montag durchführen.

Reply:
Hat zwar nichts mit diesem PR zu tun, aber könntest du das mal bitte korrigieren:

**Automaitc** updated controls and CHANGED